src/server/master_server/chunk_server_manager.h

```diff
@@ -54,45 +137,132 @@ class ChunkServerManager {
+  // mutex for threadsafe access. Returns a constant so it can't be modified
+  // outside the manager. Returns nullptr if not registered. This is used by the
+  // ChunkServerManager services only.
+  // TODO(bmokutub): Consider making this private and making the service class
+  // a friend.
+  std::pair<std::shared_ptr<const protos::ChunkServer>,
+            std::shared_ptr<absl::Mutex>>
+  GetChunkServer(const protos::ChunkServerLocation& server_location);
+
+  // Update information about a registered chunkserver. We only allow updating
+  // the available disk and chunks. Location can't be updated since it uniquely
+  // identifies this chunkserver. To change the chunkserver location, it needs
+  // to unregistered and reregistered with a new location. This is only used by
+  // the chunkserver manager service to update the chunkserver when new
+  // information is received from it. It is threadsafe.
+  // TODO(bmokutub): Consider making this private and making the service class
+  // a friend.
+  void UpdateChunkServer(
+      const protos::ChunkServerLocation& server_location,
+      const uint32_t& available_disk_mb,
+      const absl::flat_hash_set<std::string>& chunks_to_add,
+      const absl::flat_hash_set<std::string>& chunks_to_remove);
 
-  // TODO(bmokutub): Use the heartbeat to update ChunkServers info.
-  void SyncChunkServers();
+  // Returns the ChunkServerLocations for a given chunk handle.
+  // This returns all the chunk server locations where this chunk is stored.
+  // This returns an empty set if there is no location. This can be because the
+  // chunk hasn't been allocated by calling the AllocateChunkServer (dev bug) or
+  // because the chunk servers where this chunk was allocated have been
+  // unregistered (due to server down). The returned locations might be smaller
+  // than the number of allocated servers if some of those servers have been
+  // unregistered (due to server down). The ChunkServer heartbeat monitor,
+  // checks when a server is down and unregister it. If the server comes back
+  // up, it will be reregistered by the chunkservermanager service.
+  ChunkServerLocationThreadSafeFlatSet GetChunkLocations(
+      const std::string& chunk_handle);
 
  private:
-  // TODO(bmokutub): Consider a thread safe map or locking this.
-  // May not need to use shared_ptr since proto manage the memory?
-  absl::flat_hash_map<protos::ChunkServerLocation,
-                      std::shared_ptr<protos::ChunkServer>,
-                      ChunkServerLocationHash>
+  // This is a thread safe map that stores the mapping from chunkserverlocation
+  // to the chunkserver. The location uniquely identifies the chunkserver. This
+  // is used to store the registered chunk servers and are removed when the
+  // chunkserver is unregistered.
+  gfs::common::thread_safe_flat_hash_map<protos::ChunkServerLocation,
+                                         std::shared_ptr<protos::ChunkServer>,
```

Comment:
I think in generall we can replace std::share_ptr<protos::ChunkServer> by just protos::ChunkServer in this change, because this type is not expensive in terms of copy behavior. This can simplify resource management quite a bit in my opinion.

Reply:
The reason why I am using a shared_ptr instead of having copies everywhere in this code, is because the chunkserver is updated frequently in couple of places, like during allocation and update calls. That is why we also have a lock for each chunkserver. If you don't use a pointer, then during update, instead of just updating one chunkserver, you end up updating the several copies too.